Migrate table to cue model

mdvictor opened the pull request:
<!--

Thank you for sending a pull request! Here are some tips:

1. If this is your first time, please read our contribution guide at https://github.com/grafana/grafana/blob/main/CONTRIBUTING.md

2. Ensure you include and run the appropriate tests as part of your Pull Request.

3. In a new feature or configuration option, an update to the documentation is necessary. Everything related to the documentation is under the docs folder in the root of the repository.

4. If the Pull Request is a work in progress, make use of GitHub's "Draft PR" feature and mark it as such.

5. If you can not merge your Pull Request due to a merge conflict, Rebase it. This gets it in sync with the main branch.

6. Name your PR as "<FeatureArea>: Describe your change", e.g. Alerting: Prevent race condition. If it's a fix or feature relevant for the changelog describe the user impact in the title. The PR title is used to auto-generate the changelog for issues marked with the "add to changelog" label.

7. If your PR content should be added to the What's New document for the next major or minor release, add the **add to what's new** label to your PR. Note that you should add this label to the main PR that introduces the feature; do not add this label to smaller PRs for the feature.

-->

**What is this feature?**

Migrate table panel types to cue model

**Why do we need this feature?**

As part of the ongoing migration to autogenerated panel types.

**Who is this feature for?**

Everyone

**Which issue(s) does this PR fix?**:

<!--

- Automatically closes linked issue when the Pull Request is merged.

Usage: "Fixes #<issue number>", or "Fixes (paste link of issue)"

-->

Fixes #53581

**Special notes for your reviewer**:

mdvictor committed:
WIP table cue model

mdvictor committed:
WIP table types migration

ryantxu commented on packages/grafana-schema/src/common/table.cue:
```diff
@@ -17,6 +17,15 @@ TableSortByFieldState: {
 } @cuetsy(kind="interface")
 
+// TODO docs
+TableFooterCalc: {
```
This name feels a bit awkward -- maybe `TableFooterOptions`?

sdboyer commented on packages/grafana-schema/src/common/table.cue:
```diff
@@ -61,7 +70,7 @@ TableFieldOptions: {
 	// This field is deprecated in favor of using cellOptions
 	displayMode?: TableCellDisplayMode
-	cellOptions: TableCellOptions
+	cellOptions: TableCellOptions | *{ type: TableCellDisplayMode | *"auto" }
```
This is an odd construct - are you trying to work around https://github.com/grafana/cuetsy/issues/74 ?

What i'd expect here is really this:
```cue
cellOptions: TableCellOptions | *{ type: TableCellDisplayMode & "auto" }
```

but that's what doesn't work, per that other issue

mdvictor committed:
refactor

mdvictor committed:
Merge branch 'main' into mdvictor/table-cue-model

mdvictor committed:
WIP table cue

mdvictor commented:
Renamed `TableFooterCalc` to `TableFooterOptions` in table component. 

Corrected the `cellOptions` cue code.

Merged main and got the latest changes: `panelcfg.cue` and `panelcfg.gen.ts` naming

ivanortegaalba commented on packages/grafana-schema/src/common/table.cue:
```diff
@@ -17,6 +17,15 @@ TableSortByFieldState: {
 	desc?:       bool
 } @cuetsy(kind="interface")
 
+// TODO docs
```
If we want to mark this field as `maturity:experimental` we should add this docs :)

Clarity-89 replied:
Yeah, was just going to say that it'd be great if each field in this schema is documented.

mdvictor committed:
docs

mdvictor commented:
I think that this will need to be put on hold until https://github.com/grafana/cuetsy/issues/74 is fixed. I think that's the only thing stopping this PR from moving forward.

Clarity-89 commented on packages/grafana-schema/src/common/table.cue:
```diff
@@ -11,12 +11,21 @@ TableCellDisplayMode: "auto" | "color-text" | "color-background" | "color-backgr
 
+// Footer options
+TableFooterOptions: {
+	show: bool
```
Would it be possible to document each of these fields? (same for `TableSortByFieldState` and other types). The general rule is that each type property should have a comment describing what it does (even if it's obvious from the name). This will be generated as docs and we don't want them to have empty descriptions, especially considering they will be available for external use.

Clarity-89 commented:
> I think that this will need to be put on hold until [grafana/cuetsy#74](https://github.com/grafana/cuetsy/issues/74) is fixed. I think that's the only thing stopping this PR from moving forward.

I'm not sure when that issue is gonna be fixed or how easy it is to fix, but since it's the only one blocking this PR, I suggest manually fixing `defaultTableFieldOptions` in [common.types.ts](https://github.com/grafana/grafana/blob/main/packages/grafana-schema/src/veneer/common.types.ts):
```js
// TODO remove when https://github.com/grafana/cuetsy/issues/74 is fixed
export const defaultTableFieldOptions = {
  ...raw.defaultTableFieldOptions,
  cellOptions: {
    type: raw.TableCellDisplayMode.Auto,
  },
};
``` 
Also we'll need to remove the default in `table.cue`: 
```
cellOptions: TableCellOptions
```

mdvictor committed:
Merge branch 'main' into mdvictor/table-cue-model

mdvictor committed:
veneer fix, docs

mdvictor commented:
Hey @Clarity-89. Thank you for your input on this one! :D
I modified the code with the veneer bit and added some docs.
Q: There are some fields in `PanelOptions` that aren't used in the UI per se, but seem to be used in the table panel code. Would it be okay to leave those undocumented for now since they have no impact on UI/UX? I'm not sure if they are used somehow (I would assume they are though) or just remnants.

Clarity-89 reviewed (approved):
Great work! 🎉 

> Q: There are some fields in PanelOptions that aren't used in the UI per se, but seem to be used in the table panel code. Would it be okay to leave those undocumented for now since they have no impact on UI/UX? I'm not sure if they are used somehow (I would assume they are though) or just remnants.

Yeah, it's fine to document those later (or better figure out if they're still used).

ivanortegaalba reviewed (approved):
This looks great! 👍

ivanortegaalba reviewed (commented):
Just a last thing. I see there are some TODOs for docs. If we're strict with the definition of maturity `experimental`. This cannot be considered `experimental` until all docs are written

mdvictor commented:
Will have another look over the docs and fix them! Thank you!

Clarity-89 commented:
> Just a last thing. I see there are some TODOs for docs. If we're strict with the definition of maturity `experimental`. This cannot be considered `experimental` until all docs are written

I think having docs is a hard requirement for stable, not experimental.

mdvictor committed:
docs

mdvictor commented:
Hey @ivanortegaalba . I've updated the missing bits of docs related to table. Is there any other place I might have missed? Thank you

baldm0mma reviewed (approved):
amazing!
